File: solutions/tuple.dart
// tuple a b f = f a b
// fst t == a
// snd t == b

import 'church_bool.dart';

var Tuple = (a) => (b) => (f) => f(a)(b);

extension TupleDotAccess on Function {
  get fst => this(True);
  get snd => this(False);
}

void main() {
  var t = Tuple(1)(2);

  assert(t.fst == 1);
  assert(t.snd == 2);
}

File: hello-hs/src/Data/Tuple.hs
module Data.Tuple where

import           Prelude (Int, String)
data Tuple a b = Tuple a b

tuple :: a -> b -> Tuple a b
tuple = Tuple

t :: Tuple Int String
t = tuple 1 "abc"

fst :: Tuple a b -> a
fst (Tuple a _) = a
-- >>> Data.Tuple.fst t
-- 1

snd :: Tuple a b -> b
snd (Tuple _ b) = b
-- >>> Data.Tuple.snd t
-- "abc"
